test/other/testMemPoolTypes.cpp

```diff
@@ -0,0 +1,195 @@
+        counter(++counter) { }
+    ~TestData() {
+        --counter;
+    }
+    uint64_t& counter;
+};
+
+template <typename T>
+struct AllocFixture {
+    AllocFixture() : mempool(std::make_unique<T>()) {}
+
+    TestData* alloc() {
+        return mempool->template alloc<TestData>(instanceCounter);
+    }
+
+    void free(TestData* obj) {
+        return mempool->free(obj);
+    }
+
+    std::unique_ptr<T> mempool;
+    uint64_t instanceCounter=0;
+};
+
+TEMPLATE_TEST_CASE_METHOD(AllocFixture, "test allocator when full", "[class][template]",
+        (StaticMemPool<1,sizeof(TestData)>),
+        (FixedDynMemPool<sizeof(TestData), 1, 1>))
+{
+    AllocFixture<TestType> inst;
+    auto& mempool = *inst.mempool;
+    auto* obj = inst.alloc();
+    REQUIRE(obj);
+    REQUIRE(!mempool.can_alloc());
+
+    // StaticMemPool silently fails
+    //auto* obj2 = mempool.allocMem(1);
+    //REQUIRE(!obj2);
+}
+
+TEMPLATE_TEST_CASE_METHOD(AllocFixture, "test bounded allocator base functionality", "[class][template]",
+        (StaticMemPool<1024,sizeof(TestData)>),
+        (FixedDynMemPool<sizeof(TestData), 1, 1024>))
+{
+    AllocFixture<TestType> inst;
+    auto& mempool = *inst.mempool;
+
+    REQUIRE(mempool.can_alloc());
+
+    std::vector<TestData*> allocated;
+
+    for (size_t i =0; i < mempool.NUM_PAGES(); ++i) {
+        REQUIRE(mempool.can_alloc());
+        auto obj = inst.alloc();
+        REQUIRE(mempool.alloced(obj));
+        allocated.push_back(obj);
+    }
+
+    REQUIRE(inst.instanceCounter == mempool.NUM_PAGES());
+
+    REQUIRE(!mempool.can_alloc());
+
+    for (auto* obj : allocated) {
+        REQUIRE(mempool.can_free());
+        mempool.free(obj);
+    }
+    REQUIRE(inst.instanceCounter == 0);
+    REQUIRE(!mempool.can_free());
+}
+
+TEMPLATE_TEST_CASE_METHOD(AllocFixture, "test unbounded allocator base functionality", "[class][template]",
+        (DynMemPool<sizeof(TestData)>))
+{
+    AllocFixture<TestType> inst;
+    const size_t test_size = 1024;
+    auto& mempool = *inst.mempool;
+
+    REQUIRE(mempool.can_alloc());
+
+    std::vector<TestData*> allocated;
+
+    for (size_t i =0; i < test_size; ++i) {
+        REQUIRE(mempool.can_alloc());
+        auto obj = inst.alloc();
+        REQUIRE(mempool.alloced(obj));
+        allocated.push_back(obj);
+    }
+
+    REQUIRE(inst.instanceCounter == test_size);
+
+    REQUIRE(mempool.can_alloc()); // unbounded allocator can always alloc
+
+    for (auto* obj : allocated	) {
+        REQUIRE(mempool.can_free());
+        mempool.free(obj);
+    }
+    REQUIRE(inst.instanceCounter == 0);
+    REQUIRE(!mempool.can_free());
+}
+
+TEMPLATE_TEST_CASE_METHOD(AllocFixture, "test reuse allocator's memory", "[class][template]",
+        (StaticMemPool<1024,sizeof(TestData)>),
+        (FixedDynMemPool<sizeof(TestData), 1, 1024>),
+        (DynMemPool<sizeof(TestData)>))
+{
+    AllocFixture<TestType> inst;
+    auto& mempool = *inst.mempool;
+    const size_t test_size = 1024;
```

Comment:
Does this have to be the same as the values in lines 113 or 114? If yes then consider putting it outside as constexpr and using it in the template

Reply:
fixed